src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1140,32 +1190,33 @@ private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancell
         }
 
         static SolutionCompilationState ComputeFrozenPartialState(
-            SolutionCompilationState @this,
+            SolutionCompilationState frozenCompilationState,
             ArrayBuilder<DocumentState> documentStates,
             CancellationToken cancellationToken)
         {
-            var newIdToProjectStateMap = @this.SolutionState.ProjectStates;
-            var newIdToTrackerMap = @this._projectIdToTrackerMap;
-
-            foreach (var docState in documentStates)
+            var currentState = frozenCompilationState;
+            foreach (var newDocumentState in documentStates)
             {
-                // if we don't have one or it is stale, create a new partial solution
-                var tracker = @this.GetCompilationTracker(docState.Id.ProjectId);
-                var newTracker = tracker.FreezePartialStateWithDocument(docState, cancellationToken);
+                var documentId = newDocumentState.Id;
+                var oldProjectState = currentState.SolutionState.GetRequiredProjectState(documentId.ProjectId);
+                var oldDocumentState = oldProjectState.DocumentStates.GetState(documentId);
 
-                Contract.ThrowIfFalse(newIdToProjectStateMap.ContainsKey(docState.Id.ProjectId));
-                newIdToProjectStateMap = newIdToProjectStateMap.SetItem(docState.Id.ProjectId, newTracker.ProjectState);
-                newIdToTrackerMap = newIdToTrackerMap.SetItem(docState.Id.ProjectId, newTracker);
+                if (oldDocumentState is null)
+                {
+                    // Project doesn't have this document, attempt to fork it with the document added.
+                    currentState = currentState.AddDocumentsToMultipleProjects(
+                        [(oldProjectState, [newDocumentState])],
+                        static (oldProjectState, newDocumentStates)
+                        => (oldProjectState.AddDocuments(newDocumentStates), new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction(newDocumentStates)));
+                }
+                else
+                {
+                    // Project has this document, attempt to fork it with the new contents.
+                    currentState = currentState.WithDocumentState(newDocumentState);
+                }
             }
```

Comment:
this is the meat of the change.  instead of doing this update/add deep in a compilation tracker.  it is just lifted to a normal operation at the SolutionCompilationState level.